src/engine/IndexScan.cpp

```diff
@@ -562,22 +562,22 @@ struct IndexScan::SharedGeneratorState {
       // `newBlocks` or can never match any entry that is larger than the
       // entries in `joinColumn` and thus can be ignored from now on.
       metaBlocks_.removePrefix(numBlocksCompletelyHandled);
-      if (newBlocks.empty()) {
-        // The current input table matches no blocks, so we don't have to yield
-        // it.
-        continue;
+      if (!newBlocks.empty()) {
+        lastId_ = CompressedRelationReader::getRelevantIdFromTriple(
+            newBlocks.back().lastTriple_, metaBlocks_);
+      } else if (joinColumn[0] > lastId_.value_or(Id::makeUndefined())) {
+        if (!metaBlocks_.blockMetadata_.empty()) {
+          // The current input table matches no blocks, so we don't have to
+          // yield it.
+          continue;
+        }
+        // In this case there's no more data to fetch, so we no longer need to
+        // process the rest of the join side.
+        doneFetching_ = true;
+        return;
```

Comment:
This is the optimisation that prevents full consumption of the join side if possible (3).